lib/userMemoryStore.js

```diff
@@ -96,13 +98,37 @@ exports.updatePassword = function(username, oldPwd, newPwd, cb) {
   });
 };
 
+function calculateDelay(attempts) {
+  return (Math.pow(2, attempts) - 1) * ATTEMPT_DELAY_MS;
+}
+
+function doVerify(password, user, cb) {
+  // user.password here should be the stored hash
+  hash.verify(password, user.password, function(err, match) {
+    if (err) {
+      return cb(err);
+    }
+    if (!match) {
+      user.passwordAttempts = user.passwordAttempts || 0;
+      user.passwordAttempts++;
+    } else {
+      user.passwordAttempts = 0;
+    }
+    cb(null, match);
+  });
+}
+
 exports.verifyPassword = function(username, password, cb) {
   const user = findByUsername(username);
   if (!user) {
     return cb(new Error('User not found'));
```

Comment:
@paolobueno I think it would be better to always return a generic 'bad username or password' to avoid attacker know he used a wrong username and log the 'bad user' error together with the passed in user.
WDYT ?